velox/common/base/SimdUtil-inl.h

```diff
@@ -1436,4 +1436,172 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
   return true;
 }
 
+namespace detail {
+
+/// NOTE: SSE_4_2`s the performance of simdStrStr is a little slower than
+/// std::find in first-char-unmatch(read only one char per match.) Use AVX2 the
+/// performance will be better than std::find in that case.
+#if XSIMD_WITH_AVX2
+using CharVector = xsimd::batch<uint8_t, xsimd::avx2>;
+#define VELOX_SIMD_STRSTR 1
+#elif XSIMD_WITH_NEON
+using CharVector = xsimd::batch<uint8_t, xsimd::neon>;
+#define VELOX_SIMD_STRSTR 1
+#else
+#define VELOX_SIMD_STRSTR 0
+#endif
+
+extern const int kPageSize;
+
+FOLLY_ALWAYS_INLINE bool pageSafe(const void* const ptr) {
+  return ((kPageSize - 1) & reinterpret_cast<std::uintptr_t>(ptr)) <=
+      kPageSize - CharVector::size;
+}
+
+template <bool compiled, size_t kNeedleSize>
+size_t FOLLY_ALWAYS_INLINE smidStrstrMemcmp(
+    const char* s,
+    size_t n,
+    const char* needle,
+    size_t needleSize) {
+  static_assert(kNeedleSize >= 2);
+  VELOX_DCHECK_GT(needleSize, 1);
+  VELOX_DCHECK_GT(n, 0);
+  auto first = CharVector::broadcast(needle[0]);
+  auto last = CharVector::broadcast(needle[needleSize - 1]);
+  size_t i = 0;
+  // Fast path for page-safe data.
+  // It`s safe to over-read CharVector if all-data are in same page.
+  // see: https://mudongliang.github.io/x86/html/file_module_x86_id_208.html
+  // While executing in 16-bit addressing mode, a linear address for a 128-bit
+  // data access that overlaps the end of a 16-bit segment is not allowed and is
+  // defined as reserved behavior. A specific processor implementation may or
+  // may not generate a general-protection exception (#GP) in this situation,
+  // and the address that spans the end of the segment may or may not wrap
+  // around to the beginning of the segment.
+  for (; i <= n - needleSize && pageSafe(s + i) &&
```

Comment:
If I understand this correctly, we will drop out of this fast path as soon as [s + i, s + i + needleSize) across the page boundary?  So data in the subsequent pages would still go through slow path, even they are not crossing page boundary anymore?

Reply:
> So data in the subsequent pages would still go through slow path, even they are not crossing page boundary anymore

yes , your understand correctly, maybe better than that put check outside because we could still check part of data.

Reply:
But I think we can still use the fast path for the second, and any subsequent pages?  Why do we need to drop out early?

Reply:
I will rollback it to outside loop boy 

Let`s detail the case here:

![image](https://github.com/user-attachments/assets/fb40f629-0039-4d58-9825-16d1e0b5ffa6)

ref : Computer Systems: A Programmer's Perspective

>Four virtual pages (VP 1, VP 2, VP 4, and VP 7)
are currently cached in DRAM. Two pages (VP 0 and VP 5) have not yet been
allocated, and the rest (VP 3 and VP 6) have been allocated, but are not currently
cached.

I think we only need to avoid to load data at the end of VP 4  because VP 5 is not allocated yet, so we only need to check  if the end of input string array  is pageSafe

> Will we miss the opportunity if it is a super long string that is spanning multiple pages?

I think we only miss the case that the input string is right at the end of VP4

Reply:
@Yuhta  update comment and code .

Reply:
Yeah that's very clear now, thanks